File: python/1d-dynamic-programming/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        maxstring = ""
        for i in range(len(s)):
            distance, current = 0, s[i:i + 1]
            while current == current[::-1]:
                if i - distance < 0 or i + distance == len(s):
                    break
                distance += 1
                current = s[i - distance:i + distance + 1]
            distance -= 1

            if distance * 2 + 1 > len(maxstring):
                maxstring = s[i - distance:i + distance + 1]

        for i in range(len(s) - 1):
            distance, current = 0, s[i:i + 2]
            while current == current[::-1]:
                if i - distance < 0 or i + distance == len(s):
                    break
                distance += 1
                current = s[i - distance:i + distance + 2]
            distance -= 1

            if distance * 2 + 2 > len(maxstring):
                maxstring = s[i - distance:i + distance + 2]

        return maxstring

File: c++/1d-dynamic-programming/longest-palindromic-substring.cc
class Solution {
public:
    string longestPalindrome(string s) {
        std::string longest_palindrome = "";

        for (int i = 0; i < s.size(); ++i) {
            int size = 0;
            while (i - size >= 0 && i + size < s.size() && s[i - size] == s[i + size]) {
                if (size * 2 + 1 > longest_palindrome.size()) {
                    longest_palindrome = s.substr(i - size, 2 * size + 1);
                }
                size += 1;
            }

            size = 0;
            while (i - size >= 0 && i + size + 1 < s.size() && s[i - size] == s[i + size + 1]) {
                if (size * 2 + 2 > longest_palindrome.size()) {
                    longest_palindrome = s.substr(i - size, 2 * size + 2);
                }
                size += 1;
            }
        }

        return longest_palindrome;
    }
};
